### PR TITLE
improve packaging

### PR DESCRIPTION
**Pull request by dodgejoel**
this gets rid of the Makefile in lieu of direct installation with pip. I believe this is a good step towards publishing this package on pypi. . . although I have a couple of other concerns first. :D

### CODE REVIEW

**Comment by hlin117** [comment]
Hey! Thanks for contributing to my project. It seems like you have a lot of changes in this PR, including

* Moving code from `.` to `./mdlp/`
* Changing the installation process
* Changing actual code logic

Let's split these changes into three parts. This makes the code review process easier, and it lets us roll back changes when necessary.

File: mdlp/_mdlp.pyx [hunk 1]
```diff
@@ -1,16 +1,26 @@
 from __future__ import division
-from cpython.mem cimport PyMem_Malloc
-from cpython.mem cimport PyMem_Free
 import numpy as np
 cimport numpy as np
-from scipy.stats import entropy
+
+from cpython.mem cimport PyMem_Free
```
Comment (hlin117) [inline mdlp/_mdlp.pyx:6]: Why did you change the import order?

**Review by hlin117 — CHANGES_REQUESTED** [review]
It would be great if you could split these changes up into multiple parts. This makes the PR easier to review, and easier to revert if there are breaking changes.

I'm looking forward to the changes!

File: mdlp/_mdlp.pyx (hunk 1 above)
Comment (dodgejoel) [reply mdlp/_mdlp.pyx:6]: I placed the imports in alphabetical order. I don't think this is part of PEP8 explicitly but I believe is good style. In fact I meant to move the cython import above the block of 'from' imports as well to the 'import' block.

**Comment by dodgejoel** [comment]
I disagree that there are multiple parts to these changes. All code changes in this PR are in service of the core changes to setup.py. Separating out the purely formal changes to the business logic in _mdlp.pyx and _mdlp.pxd and the moving of files to the mdlp subpackage would only introduce other problems which are themselves pains to solve (e.g. get make build process to work with intermediate state of files).

It does not really make sense to make these changes separately. They are largely intertwined and should be reviewed and merged together.

File: mdlp/_mdlp.pyx (hunk 1 above)
Comment (hlin117) [reply mdlp/_mdlp.pyx:6]: Fair. Your'e right, it is good style to alphabetize imports. Thanks 👍

File: setup.py [hunk 2]
```diff
@@ -1,8 +1,53 @@
...
+      name='mdlp-discretization',
+      version='0.1',
+      description="""Implements the MDLP discretization criterion from Usama Fayyad's paper "Multi-Interval Discretization of Continuous-Valued Attributes for Classification Learning".""",
+      license='BSD 3 Clause',
+      url='github.com/hlin117/mdlp-discretization',
+      author='Henry Lin',
+      author_email='hlin117@gmail.com',
+      install_requires=[
+        'setuptools>=18.0',
+        # 'cython>=0.23.2',
```
Comment (hlin117) [inline setup.py]: Let's remove this comment.

File: setup.py [hunk 3]
```diff
@@ -1,8 +1,53 @@
...
-setup(
-    ext_modules=cythonize("*.pyx", language="c++"),
-    include_dirs=[numpy.get_include()]
-)
+from setuptools import Extension, find_packages, setup
+
+if __name__ == '__main__':
+  # see https://stackoverflow.com/questions/2379898/make-distutils-look-for-numpy-header-files-in-the-correct-place
+  # for the approach to pushing numpy an cython dependencies into extension building only
+  try:
```
Comment (hlin117) [inline setup.py]: Are these two space tab sizes? Let's make these four spaces.

File: setup.py [hunk 4]
```diff
@@ -1,8 +1,53 @@
...
+    # if Cython is available, we will rebuild from the pyx file directly
+    from Cython.Distutils import build_ext
+    sources = ['mdlp/_mdlp.pyx']
+  except:
+    # else we build from the cpp file included in the distribution
+    from setuptools.command.build_ext import build_ext
+    sources = ['mdlp/_mdlp.cpp']
+
+  class custom_build_ext(build_ext):
```
Comment (hlin117) [inline setup.py]: Rename to `CustomBuildExt`

File: setup.py [hunk 5]
```diff
@@ -1,8 +1,53 @@
...
-import numpy
-
-setup(
-    ext_modules=cythonize("*.pyx", language="c++"),
-    include_dirs=[numpy.get_include()]
-)
+from setuptools import Extension, find_packages, setup
+
+if __name__ == '__main__':
+  # see https://stackoverflow.com/questions/2379898/make-distutils-look-for-numpy-header-files-in-the-correct-place
```
Comment (hlin117) [inline setup.py]: Thanks for the stackoverflow link. I'm a little bothered that it's more than 72 characters long though.

You can probably get away with using `https://stackoverflow.com/a/42163080/2014591`.

File: setup.py [hunk 6]
```diff
@@ -1,8 +1,53 @@
...
+  # for the approach to pushing numpy an cython dependencies into extension building only
```
Comment (hlin117) [inline setup.py]: 72 characters long, please.

Maybe I should add a lint checker to this project sometime...

File: setup.py [hunk 7]
```diff
@@ -1,8 +1,53 @@
...
+    from Cython.Distutils import build_ext
+    sources = ['mdlp/_mdlp.pyx']
+  except:
+    # else we build from the cpp file included in the distribution
+    from setuptools.command.build_ext import build_ext
+    sources = ['mdlp/_mdlp.cpp']
+
+  class custom_build_ext(build_ext):
+    """
+    custom class to add numpy include dirs when we are building the extension.
```
Comment (hlin117) [inline setup.py]: Please read the PEP 8 recommendations for docstrings here: https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

File: _mdlp.pxd [hunk 8]
```diff
@@ -1,21 +0,0 @@
-import numpy as np
```
Comment (hlin117) [inline _mdlp.pxd:1]: I'm still concerned that you put in extra changes in this PR to remove this file, and move its contents to `mdlp/_mdlp.pyx`. But overall, the changes are good, and you were able to remove some dead code here.

**Review by hlin117 — COMMENTED** [review]
Overall, the changes look great. Thanks for your time in making this project better.

File: _mdlp.pxd (hunk 8 above)
Comment (dodgejoel) [reply _mdlp.pxd:1]: Certainly your concern makes sense. If I were more familiar with github pr workflows I would be more aggressively bending to your request to separate things.

File: setup.py (hunk 5 above)
Comment (dodgejoel) [reply setup.py]: Oh lovely thank you. I had never used the 'share' button on SO before. TIL!

File: setup.py (hunk 4 above)
Comment (dodgejoel) [reply setup.py]: done.

File: setup.py (hunk 2 above)
Comment (dodgejoel) [reply setup.py]: done.

File: setup.py (hunk 6 above)
Comment (dodgejoel) [reply setup.py]: Sounds good!

File: setup.py (hunk 3 above)
Comment (dodgejoel) [reply setup.py]: Ah of course. I did not think to match the style in the package and I typically use 2 space tabs. Sorry about that!

File: setup.py (hunk 7 above)
Comment (dodgejoel) [reply setup.py]: great thank you.

**Comment by dodgejoel** [comment]
It looks like you are in the process of replacing travis with AppVeyor or something? I guess LMK when your dust clears and I will make sure the build is working.

**Comment by hlin117** [comment]
> It looks like you are in the process of replacing travis with AppVeyor or something? I guess LMK when your dust clears and I will make sure the build is working.

Interesting. I didn't know AppVeyor would override Travis. Please make an empty commit

```
git commit --allow-empty -m "Retrying build"
```

And try getting the build to be green. Thanks!

**Comment by hlin117** [comment]
Theoretically a build should be able to run Travis and AppVeyor. AppVeyor is for Windows, while Travis is for Unix.

File: README.md [hunk 9]
```diff
@@ -9,45 +9,46 @@ for more information.
...
-and then run
+As with all python packages, it is recommended to create a virtual environment
+when using this project.
+
+# Example
 
 ```
-make
-```
+from mdlp.discretization import MDLP
```
Comment (hlin117) [inline README.md]: Please put back the `>>>`. This allow python doctests to be ran
https://docs.python.org/2/library/doctest.html

**Review by hlin117 — CHANGES_REQUESTED** [review]

**Comment by dodgejoel** [comment]
It looks like the travis integration has been disabled or something. At least the build is not triggered on push as it previously was for this repo.

**Comment by hlin117** [comment]
Hmm,

1. According to [here](https://github.com/dodgejoel/mdlp-discretization/network) I don't think you rebased your changes off of master, which is why there are merge conflicts. (Or, at least the repository is detecting merge conflicts.) This may explain why the build is not kicking off. 
2. Manually running the build, it looks like Travis is green. See [here](https://travis-ci.org/hlin117/mdlp-discretization/builds/290898131).

**Comment by hlin117** [comment]
It looks like Travis usually doesn't kick off unless the project is mergable. See [here](https://github.com/travis-ci/travis-ci/issues/4102#issuecomment-112500712).

**Comment by dodgejoel** [comment]
I see that all makes sense. I have gotten git into somewhat of a messy bind. I will push the accumulated changes to a new branch and issue a new pull request that should hopefully get things working.

**Comment by hlin117** [comment]
For future reference, you should rebase, and then `git push -f`. [This might help](https://github.com/servo/servo/wiki/Beginner's-guide-to-rebasing-and-squashing). 

After a while of using git, rebasing / squashing becomes less scary, and more of a natural technique :)
